BeVolt/App/src/CLI.c

```diff
@@ -0,0 +1,723 @@
+		EEPROM_SerialPrintData();
+		return;
+	}
+	uint8_t errorAddrArray[2];
+	uint16_t errorAddr = 0;
+	switch(hashTokens[1]) {
+		case RESET:
+			EEPROM_Reset();
+			printf("EEPROM has been reset");
+			break;
+		case ERROR:
+			switch(hashTokens[2]) {
+				case FAULT:
+					EEPROM_ReadMultipleBytes(EEPROM_FAULT_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("Fault error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				case TEMPERATURE:
+					EEPROM_ReadMultipleBytes(EEPROM_TEMP_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("Temperature error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				case VOLTAGE:
+					EEPROM_ReadMultipleBytes(EEPROM_VOLT_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("Voltage error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				case CURRENT:
+					EEPROM_ReadMultipleBytes(EEPROM_CURRENT_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("Current error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				case WATCHDOG:
+					EEPROM_ReadMultipleBytes(EEPROM_WATCHDOG_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("Watchdog error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				case CAN:
+					EEPROM_ReadMultipleBytes(EEPROM_CAN_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("CAN error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				case CHARGE:
+					EEPROM_ReadMultipleBytes(EEPROM_SOC_PTR_LOC, 2, errorAddrArray);
+					errorAddr = (errorAddrArray[0] << 2) + errorAddrArray[1];
+					printf("SoC error: %d", EEPROM_ReadByte(errorAddr-1));
+					break;
+				default:
+					break;
+			}
+			break;
+		default:
+			printf("Invalid EEPROM command\n\r");
+			break;
+	}
+}
+
+/** CLI_ADC
+ * Prints the high precision and
+ * low precision readings
+ */
+void CLI_ADC(void) {
+	printf("High precision ADC: %d\n\r", ADC_ReadHigh());
+	printf("Low precision ADC: %d\n\r", ADC_ReadLow());
+}
+
+
+/** CLI_Critical
+ * Shuts off contactor manually
+ */  
+void CLI_Critical(void) {
+	Fifo criticalFifo;
+	char response[fifo_size];
+	printf("Please type 'shutdown' to turn the contactor off\n\r");
+	printf(">> ");
+	while(1) {
+		UART3_CheckAndEcho(&criticalFifo);
+		if(UART3_HasCommand(&criticalFifo)) {
+			UART3_GetCommand(&criticalFifo, response);
+			if(CLI_StringHash(response) == SHUTDOWN) {
+				Contactor_Off();
+				printf("Contactor is off\n\r");
+				break;
+			} else {
+				printf("Contactor is still on\n\r");
+				break;
+			}
+		}
+	}
+}
+
+/** CLI_All
+ * Displays all information about BPS modules
+ * (voltage, current, temperature, charge, contactor)
+ */
+void CLI_All(void){
+	printf("Voltage: \n\r");
+	hashTokens[0] = VOLTAGE;

```

Comment:
An example of awkwardness arising from the use of a global variable.

Reply:
I changed it to a locally passed array, but I'm not sure how to set these values appropriately before calling the respective functions without doing something like this